### PR TITLE
Update redacting functionality to redact all sensitive info in config when printing with view

### LINKED ISSUE #1193: kubectl config view shows auth-provider secrets
**What happened**:
`kubectl config view` (without `--raw` flag) print `users[*].user.auth-provider.config.{client-secret, id-token, refresh-token}` values. Also command prints `users.[*].user.password` value. 

**What you expected to happen**:
I think this command should print `DATA+OMITTED` (or `REDACTED`) word like in `clusters.[*].cluster.certificate-authority-data`

**Environment**:
- Kubernetes client version (use `kubectl version`): 1.23.4 (e6c093d87ea4cbb530a7b2ae91e54c0842d8308a)
- OS: macOS Montery (12.1)

### PR DESCRIPTION
**Pull request by mpuckett159**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Updates redacting functionality of `kubectl config view` to properly redact all config keys with a `datapolicy` tag. This is accomplished by introducing a new function `RedactSecrets` in client-go and then consuming it in kubectl.

Example usages

No flags:
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config view
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config:
        refresh-token: test1233
      name: ""
    client-certificate-data: DATA+OMITTED
    client-key-data: DATA+OMITTED
    exec:
      args:
      - test
      command: foo1
      env:
      - name: test
        value: vla2
      provideClusterInfo: false
    token: REDACTED
    password: REDACTED
    username: mpuckett
```

With --raw:
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config view --raw         
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config:
        refresh-token: test1233
      name: ""
    client-certificate-data: YWZmYWRzZmFzZGZz
    client-key-data: YWZmYWRzZmFzZGZz
    exec:
      args:
      - test
      command: foo1
      env:
      - name: test
        value: vla2
      provideClusterInfo: false
    token: YWZmYWRzZmFzZGZz
    password: fake-password
    username: mpuckett
```

#### Which issue(s) this PR fixes:
Fixes kubernetes/kubectl#1193

#### Does this PR introduce a user-facing change?
```release-note
kubectl config view now automatically redacts any secret fields marked with a datapolicy tag
```

### CODE REVIEW

**Comment by mpuckett159** [comment]
You'll notice in the above examples that `users.foo.auth-provider.config.refresh-token` is not redacted. I'm not exactly sure how we should god about doing that if it is desired to redact that, because this value is set by a simple `map[string]string` which can't have the `datapolicy` that I'm using to determine whether to redact or not. We could search for map keys that contain `token` or `password` maybe, even `secret` but because this is kind of a free form field it might just be better to manually remove the output of the entire object.

File: staging/src/k8s.io/kubectl/pkg/cmd/config/view.go [hunk 1]
```diff
@@ -96,6 +95,7 @@ func NewCmdConfigView(streams genericclioptions.IOStreams, ConfigAccess clientcm
 	cmd.Flags().BoolVar(&o.RawByteData, "raw", o.RawByteData, "Display raw byte data")
 	cmd.Flags().BoolVar(&o.Flatten, "flatten", o.Flatten, "Flatten the resulting kubeconfig file into self-contained output (useful for creating portable kubeconfig files)")
 	cmd.Flags().BoolVar(&o.Minify, "minify", o.Minify, "Remove all information not used by current-context from the output")
+	cmd.Flags().BoolVar(&o.ShowSecrets, "show-secrets", o.ShowSecrets, "Display sensitive data")
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/config/view.go:98]: Why we are introducing new flag?. As far as I understand from the issue, `--raw` will show sensitives and others will not? or none will not show any sensitive?

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go [hunk 2]
```diff
@@ -93,13 +95,10 @@ func ShortenConfig(config *Config) {
 	// http://golang.org/pkg/encoding/json/#Marshal using base64 to encode []byte
 	for key, authInfo := range config.AuthInfos {
 		if len(authInfo.ClientKeyData) > 0 {
-			authInfo.ClientKeyData = redactedBytes
+			authInfo.ClientKeyData = dataOmittedBytes
 		}
 		if len(authInfo.ClientCertificateData) > 0 {
-			authInfo.ClientCertificateData = redactedBytes
-		}
-		if len(authInfo.Token) > 0 {
```
Comment (ardaguclu) [inline staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go:101]: Why this conditional is removed?

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go [hunk 3]
```diff
@@ -93,13 +95,10 @@ func ShortenConfig(config *Config) {
 	// http://golang.org/pkg/encoding/json/#Marshal using base64 to encode []byte
 	for key, authInfo := range config.AuthInfos {
 		if len(authInfo.ClientKeyData) > 0 {
-			authInfo.ClientKeyData = redactedBytes
+			authInfo.ClientKeyData = dataOmittedBytes
```
Comment (ardaguclu) [inline staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go:98]: nit: for the simplicity of PR, maybe it would be better not renaming this variables and do this another PR?

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go [hunk 4]
```diff
@@ -189,3 +188,79 @@ func MakeAbs(path, base string) (string, error) {
...
+
+// RedactSecrets replaces any sensitive values with REDACTED
+func RedactSecrets(config *Config) error {
+	if err := redactSecrets(reflect.ValueOf(config), false); err != nil {
+		return err
+	}
+	return nil
+}
+
+func redactSecrets(curr reflect.Value, redact bool) error {
```
Comment (ardaguclu) [inline staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go:200]: As far as I can see, `datapolicy` is set only for 3 fields https://github.com/kubernetes/kubernetes/blob/7e079f5144474cae05802771f604df2b748d781f/staging/src/k8s.io/client-go/tools/clientcmd/api/types.go#L117-L120 and `Config` struct is not something very often changing.

Would it be better to redact just these ^ fields statically instead this dynamic recursive function?.

I wonder other reviewers opinion about this.

**Review by ardaguclu — COMMENTED** [review]
Thanks for the PR @mpuckett159 !

File: staging/src/k8s.io/kubectl/pkg/cmd/config/view.go (hunk 1 above)
Comment (mpuckett159) [reply staging/src/k8s.io/kubectl/pkg/cmd/config/view.go:98]: This flag was added because, due to the verbiage of the flag help text for `--raw`, it is not immediately apparent to the user that `--raw` will be displaying secrets. It seemed to me like the best option was to add a new flag to explicitly display secrets, because there may be situations where users will want to display the raw byte data of their client certificate data, but not their client key data as that is a secret. Admittedly this is a bit of a breaking change in terms of user expectation.

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go (hunk 2 above)
Comment (mpuckett159) [reply staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go:101]: This is within the ShortenConfig() function, which with this PR would no longer be in control of the token field and would instead be in charge of simply shortening the config by omitting the base64 encoded certificate data. The Token field is switched to the expected `REDACTED` output later by the RedactSecrets function if it is required by the flags the user has set.

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go (hunk 3 above)
Comment (mpuckett159) [reply staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go:98]: In the GitHub viewer it's hard to see because it removes pretty much all the context, but this line change is done in the `ShortenConfig()` function, which via this PR no longer handles the redaction of secrets as that has been offloaded to the new `redactSecrets()` function, so here for clarity I've updated the variable to be the `dataOmittedBytes`.

File: staging/src/k8s.io/kubectl/pkg/cmd/config/view.go (hunk 1 above)
Comment (mpuckett159) [reply staging/src/k8s.io/kubectl/pkg/cmd/config/view.go:98]: Do we think there will be a path forward to move to using both of these flags? I think it is important to give users the option to view their raw bytes with out forcing them to also expose their secrets. I know it's a relatively small threat vector but it feels like an easy enough gap to close as long as we do it right without breaking things for people unexpectedly.

Comment (mpuckett159) [reply staging/src/k8s.io/kubectl/pkg/cmd/config/view.go:98]: Thinking about this more, wonder if we can create two new flags, `--show-bytes` and `--show-secrets`, note that `--raw` will perform the function of both in the help text as that matches the current functionality of the flag, and either plan a deprecation for `--raw` or just leave `--raw` as a way for users to be able to functionally pass both flags at once.

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go (hunk 2 above)
Comment (mpuckett159) [reply staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go:101]: Understood, I've reverted this conditional change and updated the tests to reflect this. I will update the PR description to incorporate this change and the new flag and reversion of `--raw` to it's original functionality.

**Comment by mpuckett159** [comment]
I still need to write unit tests for the new function in helpers.

File: staging/src/k8s.io/kubectl/pkg/cmd/config/view.go (hunk 1 above)
Comment (soltysh) [reply staging/src/k8s.io/kubectl/pkg/cmd/config/view.go:105]: Let's keep it simple, I bet you can easily figure out additional 2 or 3 flags for that command, but unless we'll have a direct feedback from users I'd prefer to keep it the way it is :wink:

Comment (mpuckett159) [reply staging/src/k8s.io/kubectl/pkg/cmd/config/view.go]: Alright removed new flags.

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go [hunk 5]
```diff
@@ -189,3 +191,79 @@ func MakeAbs(path, base string) (string, error) {
 	}
 	return filepath.Join(base, path), nil
 }
+
+// RedactSecrets replaces any sensitive values with REDACTED
+func RedactSecrets(config *Config) error {
+	if err := redactSecrets(reflect.ValueOf(config), false); err != nil {
```
Comment (soltysh) [inline staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go:197]: Make it simpler :wink:, you don't need the if at all:
```golang
	return redactSecrets(reflect.ValueOf(config), false)
```

**Review by soltysh — APPROVED** [review]
One minor nit, can you also squash the changes to a single commit?

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go (hunk 5 above)
Comment (mpuckett159) [reply staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go:197]: Doh, thank you.

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers_test.go [hunk 6]
```diff
@@ -293,11 +292,60 @@ func TestShortenSuccess(t *testing.T) {
...
+		t.Errorf("expected %v, got %v", startingConfig.Clusters[unchangingCluster], mutatingConfig.Clusters[unchangingCluster])
+	}
+
+	if len(mutatingConfig.AuthInfos) != 2 {
+		t.Errorf("unexpected users: %v", mutatingConfig.AuthInfos)
+	}
+	if !reflect.DeepEqual(startingConfig.AuthInfos[unchangingAuthInfo], mutatingConfig.AuthInfos[unchangingAuthInfo]) {
+		t.Errorf("expected %v, got %v", startingConfig.AuthInfos[unchangingAuthInfo], mutatingConfig.AuthInfos[unchangingAuthInfo])
 	}
 	if mutatingConfig.AuthInfos[changingAuthInfo].Token != "REDACTED" {
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/tools/clientcmd/api/helpers_test.go:350]: shouldn't the key data also be redacted?

File: staging/src/k8s.io/kubectl/pkg/cmd/config/view.go [hunk 7]
```diff
@@ -136,6 +134,12 @@ func (o ViewOptions) Run() error {
 		return err
 	}
 
+	if !o.RawByteData {
```
Comment (deads2k) [inline staging/src/k8s.io/kubectl/pkg/cmd/config/view.go:137]: I would expect this as the last possible step before printing.

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go [hunk 8]
```diff
@@ -189,3 +191,76 @@ func MakeAbs(path, base string) (string, error) {
 	}
 	return filepath.Join(base, path), nil
 }
+
+// RedactSecrets replaces any sensitive values with REDACTED
+func RedactSecrets(config *Config) error {
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go:196]: Returning a copy that is redacted instead of mutating the original seems like more obvious behavior. With generated deepcopy methods, this ought not be too bad to do.

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers_test.go (hunk 6 above)
Comment (mpuckett159) [reply staging/src/k8s.io/client-go/tools/clientcmd/api/helpers_test.go:350]: We've been having a bit of an argument about this.

Originally I had separate flags, `--show-bytes` and `--show-secrets`, so that a user could potentially show raw bytes without exposing secrets, which would allow them to show e.g. a CA cert's data, but still hide the client key data, and also show the password/token/etc fields without having to display all the cert byte data, and `--raw` just set both of those flags to true basically. It was deemed that this was too much of a departure from the original functionality and would be too much of a breaking change.

File: staging/src/k8s.io/kubectl/pkg/cmd/config/view.go (hunk 7 above)
Comment (mpuckett159) [reply staging/src/k8s.io/kubectl/pkg/cmd/config/view.go:137]: Hrm, I'm not sure I understand this. I don't think the ordering should matter really, but I just stuck this at the front because I figured it would be best to redact sensitive info as soon as possible.

Comment (soltysh) [reply staging/src/k8s.io/kubectl/pkg/cmd/config/view.go:137]: @mpuckett159 what David wants is to move this down right before `return o.PrintObject(convertedObj, o.Out)`, this way all the minifying and flattening will always take place, and most importantly conversion to external scheme. 
This is good point, i haven't considered before.

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go (hunk 8 above)
Comment (soltysh) [reply staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go:196]: I'd definitely support that rewrite in a followup PR. This will be a change in behavior of all the public functions.

File: staging/src/k8s.io/kubectl/pkg/cmd/config/view.go (hunk 7 above)
Comment (mpuckett159) [reply staging/src/k8s.io/kubectl/pkg/cmd/config/view.go:137]: Apologies, I think I'm still missing something. The helper I wrote specifically works on the clientcmdapi Config object so that it can make use of the datapolicy tags to know what to redact without relying on manually adding keys to be redacted to the functions. I don't think that option would be available when working on the runtime Object interface. I could just cast it back to a Config I assume, but that feels wrong because I'd have to cast it back to a runtime Object before returning again.

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers_test.go (hunk 6 above)
Comment (mpuckett159) [reply staging/src/k8s.io/client-go/tools/clientcmd/api/helpers_test.go:351]: Sorry just realized I missed what you were asking about. Correct, should be redacted, updated the test to include that as well. Good catch because I was not setting the client key data properly.

File: staging/src/k8s.io/kubectl/pkg/cmd/config/view.go (hunk 7 above)
Comment (soltysh) [reply staging/src/k8s.io/kubectl/pkg/cmd/config/view.go:137]: Just looked through the code again, and see where the confusion comes from, it's fine to use your method as is before the conversion and printing, so I'd suggest to ideally put it here: https://github.com/kubernetes/kubernetes/blob/9b9a963a8d0750735e2e6bbe4af5d1bb7cfc7f02/staging/src/k8s.io/kubectl/pkg/cmd/config/view.go#L152-L154

**Comment by mpuckett159** [comment]
Sorry for the churn there, not exactly sure what I was thinking there 🙄

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
/label tide/merge-method-squash

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go (hunk 8 above)
Comment (sttts) [reply staging/src/k8s.io/client-go/tools/clientcmd/api/helpers.go:195]: +1 for what David says. But yeah, this package likes in-place updates.

File: staging/src/k8s.io/client-go/tools/clientcmd/api/helpers_test.go [hunk 9]
```diff
@@ -240,8 +241,8 @@ func Example_minifyAndShorten() {
 	// users:
 	//   red-user:
 	//     LocationOfOrigin: ""
-	//     client-certificate-data: REDACTED
-	//     client-key-data: REDACTED
+	//     client-certificate-data: DATA+OMITTED
+	//     client-key-data: DATA+OMITTED
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/clientcmd/api/helpers_test.go:245]: did this really change?

Comment (sttts) [reply staging/src/k8s.io/client-go/tools/clientcmd/api/helpers_test.go:245]: and why

Comment (mpuckett159) [reply staging/src/k8s.io/client-go/tools/clientcmd/api/helpers_test.go:245]: This test is for the ShortenConfig() function which no longer handles the process of redacting secrets and instead only omits byte data.
